### PR TITLE
Add support for .tgz files in GeoIpDownloader

### PR DESCRIPTION
**Pull request by probakowski**
We have to ship `COPYRIGHT.txt` and `LICENSE.txt` files alongside `.mmdb` files for legal compliance. Infra will pack these in single `.tgz` (gzipped tar) archive provided by GeoIP databases service. 
This change adds support for that format to `GeoIpDownloader` and `DatabaseRegistry`

### CODE REVIEW

**Comment by probakowski** [comment]
@elasticmachine update branch

File: libs/core/src/main/java/org/elasticsearch/common/io/TarEntry.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.io;
+
+import java.util.Objects;
+
+public class TarEntry {
```
Comment (martijnvg) [inline libs/core/src/main/java/org/elasticsearch/common/io/TarEntry.java:13]: Maybe we should use a library that provides tar ball support? Like plexus-archiver or Apache commons-compress?

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpTaskState.java [hunk 2]
```diff
@@ -160,12 +158,18 @@ public static Metadata fromXContent(XContentParser parser) {
         private final int firstChunk;
         private final int lastChunk;
         private final String md5;
+        private final boolean tar;
```
Comment (martijnvg) [inline modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpTaskState.java:161]: Are all database files now provided as tarball? If so then we don't need this `tar` field?

File: modules/ingest-geoip/src/internalClusterTest/java/org/elasticsearch/ingest/geoip/GeoIpDownloaderIT.java [hunk 3]
```diff
@@ -254,7 +263,7 @@ public void testUseGeoIpProcessorWithDownloadedDBs() throws Exception {
             for (Path geoipTmpDir : geoipTmpDirs) {
                 try (Stream<Path> list = Files.list(geoipTmpDir)) {
                     List<String> files = list.map(Path::toString).collect(Collectors.toList());
-                    assertThat(files, containsInAnyOrder(endsWith("GeoLite2-City.mmdb"), endsWith("GeoLite2-Country.mmdb"),
+                    assertThat(files, hasItems(endsWith("GeoLite2-City.mmdb"), endsWith("GeoLite2-Country.mmdb"),
```
Comment (martijnvg) [inline modules/ingest-geoip/src/internalClusterTest/java/org/elasticsearch/ingest/geoip/GeoIpDownloaderIT.java:266]: maybe we should ensure that the text files exist as well?

**Review by martijnvg — COMMENTED** [review]
Looks good. Left a few questions.

File: libs/core/src/main/java/org/elasticsearch/common/io/TarEntry.java (hunk 1 above)
Comment (probakowski) [reply libs/core/src/main/java/org/elasticsearch/common/io/TarEntry.java:13]: I thought about it but both libraries adds quite big footprint (common-cmpress over 600kB, plexus over 200kB) just for this one use case. 
I think tar is simple enough (at least for basic cases but I don't expect anything crazy from infra) to parse it by ourselves. That said I can reconsider if you are really against creating our own solution here.

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpTaskState.java (hunk 2 above)
Comment (probakowski) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpTaskState.java:161]: Right now they provide both versions, `mmdb.gz` and `.tgz`. I can filter and process only `tgz` files, then I can remove this field. We could then remove handling of simple `.gz` altogether.

File: modules/ingest-geoip/src/internalClusterTest/java/org/elasticsearch/ingest/geoip/GeoIpDownloaderIT.java (hunk 3 above)
Comment (probakowski) [reply modules/ingest-geoip/src/internalClusterTest/java/org/elasticsearch/ingest/geoip/GeoIpDownloaderIT.java:266]: This is good idea, I'll change that

File: libs/core/src/main/java/org/elasticsearch/common/io/TarEntry.java (hunk 1 above)
Comment (martijnvg) [reply libs/core/src/main/java/org/elasticsearch/common/io/TarEntry.java:13]: I just realized that `plexus-archiver` actually depends on `common-compress`. I agree it is a lot of additional bytes for just for extracting a tarball and the code itself isn't complex. On the hand, about half of this PR is about adding support for extracting a tarball and comparing the size of all our other dependencies (all jdk libraries and the jdk) I think about 600kb for common-compress is just a small part our total distribution.

I did find a lightweight library called `jtar` that adds support for just tar files, but that project doesn't seem very active and maintained by basically a single developer.

I think I'm ok with maintaining this tar code ourselves, but it should be moved to ingest-geoipv2 module and explained that it is not a general tar library and meant to be used for extracting the tar files provided by the infra service.

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpTaskState.java (hunk 2 above)
Comment (martijnvg) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpTaskState.java:161]: That sounds good to me. This avoids having multiple code paths that essentially have the same purpose.

File: modules/ingest-geoip/src/internalClusterTest/java/org/elasticsearch/ingest/geoip/GeoIpDownloaderIT.java (hunk 3 above)
Comment (probakowski) [reply modules/ingest-geoip/src/internalClusterTest/java/org/elasticsearch/ingest/geoip/GeoIpDownloaderIT.java]: Done

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpTaskState.java (hunk 2 above)
Comment (probakowski) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpTaskState.java]: Done

File: libs/core/src/main/java/org/elasticsearch/common/io/TarEntry.java (hunk 1 above)
Comment (probakowski) [reply libs/core/src/main/java/org/elasticsearch/common/io/TarEntry.java]: I moved it to geoip module and made it simpler by skipping everything that is not needed for our use case.

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/TarInputStream.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+/**
+ * {@link InputStream} with very basic support for tar format, just enough to parse archives provided by GeoIP database service from Infra.
+ * This class is not suitable for general purpose tar processing!
+ */
+class TarInputStream extends FilterInputStream {
```
Comment (martijnvg) [inline modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/TarInputStream.java:22]: 👍

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java [hunk 5]
```diff
@@ -261,6 +262,21 @@ void retrieveAndUpdateDatabase(String databaseName, GeoIpTaskState.Metadata meta
                 decompress(databaseTmpGzFile, databaseTmpFile);
 
                 Path databaseFile = geoipTmpDirectory.resolve(databaseName);
+                try (TarInputStream is = new TarInputStream(new BufferedInputStream(Files.newInputStream(databaseTmpFile)))) {
```
Comment (martijnvg) [inline modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java:265]: maybe leave a comment here about the structure of the tar file and how it is going to be extracted?

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java [hunk 6]
```diff
@@ -261,6 +262,21 @@ void retrieveAndUpdateDatabase(String databaseName, GeoIpTaskState.Metadata meta
...
+                    TarInputStream.TarEntry entry;
+                    while ((entry = is.getNextEntry()) != null) {
+                        if (entry.isNotFile()) {
+                            continue;
+                        }
+                        String name = entry.getName().substring(entry.getName().lastIndexOf('/') + 1);
+                        if (name.startsWith(databaseName)) {
+                            Files.copy(is, databaseTmpFile, StandardCopyOption.REPLACE_EXISTING);
+                        } else {
+                            Files.copy(is, geoipTmpDirectory.resolve(databaseName + "_" + name), StandardCopyOption.REPLACE_EXISTING);
```
Comment (martijnvg) [inline modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java:275]: The `removeStaleEntries()` method doesn't remove these files, but I think that is ok.
These are small files and each update overwrites these files, during node startup the
geoip tmp dir is purged and the `removeStaleEntries()` is only invoked if at some 
point in time we stop distributing databases by default or custom dbs are shipped
by a third party database webserver.

**Review by martijnvg — APPROVED** [review]
LGTM

**Comment by probakowski** [comment]
@elasticmachine run elasticsearch-ci/1

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java (hunk 5 above)
Comment (probakowski) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java:267]: Added

Comment (probakowski) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java:267]: Added

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java (hunk 6 above)
Comment (probakowski) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java:279]: Yes, that was exactly my thinking, I think we can leave them behind as they are small and doesn't interfere with geoip processor itself.

**Pull request merged**